applications/main/clock/clock_app.c

```diff
@@ -0,0 +1,197 @@
+#include <furi.h>
+#include <furi_hal.h>
+
+#include <gui/gui.h>
+#include <gui/elements.h>
+
+#include "clock_app.h"
+
+static void clock_input_callback(InputEvent* input_event, FuriMessageQueue* event_queue) {
+    furi_assert(event_queue);
+    PluginEvent event = {.type = EventTypeKey, .input = *input_event};
+    furi_message_queue_put(event_queue, &event, FuriWaitForever);
+}
+
+static void clock_render_callback(Canvas* const canvas, void* ctx) {
+    FURI_LOG_T(TAG, "clock_render_callback - start");
+
+    canvas_clear(canvas);
+    canvas_set_color(canvas, ColorBlack);
+
+    ClockState* state = ctx;
+    if(furi_mutex_acquire(state->mutex, 200) != FuriStatusOk) {
+        FURI_LOG_D(TAG, "Can't obtain mutex, requeue render");
+        PluginEvent event = {.type = EventTypeTick};
+        furi_message_queue_put(state->event_queue, &event, 0);
```

Comment:
IIRC I wrote that last time this went around. Reason this was here was because drawing only occurs every second, so if mutex was locked by update logic, you'd miss drawing and the display would end up blank for a whole second. Maybe update loop could tick more frequently and use a flag for whether a redraw is required.

Reply:
There are 2 ways how you can solve it:

- Lock-free data model: when all fields are atomic then you don't have to lock it. For example if used types not exceed machine word.
- Representation Model filled in one go and mutex is only held for a short time. (View is doing it).

Both can be applied here.